### PR TITLE
Implements SetArchiveLayout and async/await

### PR DESCRIPTION
**Pull request by gleborgne**
SetArchiveLayout method was not implemented and required for our business. We also implements async/await to avoid IO blocking threads and improve performances.

### CODE REVIEW

File: OpenTok/OpenTok.cs [hunk 1]
```diff
@@ -142,6 +143,84 @@ public OpenTok(int apiKey, string apiSecret, string apiUrl)
         /// method (to initialize an OpenTok session).
         /// </returns>
         public Session CreateSession(string location = "", MediaMode mediaMode = MediaMode.RELAYED, ArchiveMode archiveMode = ArchiveMode.MANUAL)
+        {
```
Comment (slorello89) [inline OpenTok/OpenTok.cs:146]: For these synchronous methods, we should maintain their current pattern, I'm concerned that we're exposing ea deadlock condition here if you end up running this on a single-threaded webserver. I'm ok with keeping a method for each sync/async but they should be sync/async all the way up and down the stack.

File: OpenTok/Util/HttpClient.cs [hunk 2]
```diff
@@ -62,26 +63,63 @@ public virtual string Get(string url, Dictionary<string, string> headers)
...
         }
 
+        public virtual Task<string> DeleteAsync(string url, Dictionary<string, string> headers)
+        {
+            headers.Add("Method", "DELETE");
+            return DoRequestAsync(url, headers, null);
+        }
+
         public string DoRequest(string url, Dictionary<string, string> specificHeaders,
                                         Dictionary<string, object> bodyData)
```
Comment (slorello89) [inline OpenTok/Util/HttpClient.cs:114]: Same with the DoRequest method - would like to keep this sync down the stack.

File: OpenTokTest/OpenTokTest.cs [hunk 3]
```diff
@@ -60,7 +60,7 @@ public void CreateSimpleSessionTest()
             var expectedUrl = "session/create";
 
             var mockClient = new Mock<HttpClient>();
-            mockClient.Setup(httpClient => httpClient.Post(It.IsAny<string>(), It.IsAny<Dictionary<string, string>>(), It.IsAny<Dictionary<string, object>>())).Returns(returnString);
```
Comment (slorello89) [inline OpenTokTest/OpenTokTest.cs:63]: will probably need to do 2 versions of these tests to test each sync/async method.

**Review by slorello89 — CHANGES_REQUESTED** [review]
Hello @gleborgne - definitely a quality PR and a solid contribution - see my comments contained in here - but I do have a bit of a concern about using the Wait/Result portions of the async methods as in certain environments they can introduce deadlocks in single-threaded environments as it can block the current thread, which may be needed to finish executing the async portions of the method underneath it. 

Realistically we should be maintaining a sync/async version of all these methods and those methods should be sync or async all the way up and down the stack.

Also - there's a `OpenTokSDK.csproj` that was created somehow - not quite sure where that came from - but everything should be in OpenTok.csproi.

Thanks!

**Comment by gleborgne** [comment]
I refactor to have distinc async and non async path in the code.

**Comment by slorello89** [comment]
@gleborgne - FYI I'm updating your branch (updating the GH actions) - so make sure you do a pull before proceeding

**Review by slorello89 — CHANGES_REQUESTED** [review]

File: OpenTok/OpenTok.cs [hunk 4]
```diff
@@ -584,9 +1088,34 @@ public Broadcast StopBroadcast(string broadcastId)
         /// including the broadcast ID.
         /// </returns>
         public Broadcast GetBroadcast(string broadcastId)
+        {
+            try
+            {
+                var task = GetBroadcastAsync(broadcastId);
+                task.Wait();
```
Comment (slorello89) [inline OpenTok/OpenTok.cs]: Still using Task.Wait()

File: OpenTok/OpenTok.cs [hunk 5]
```diff
@@ -564,11 +1042,37 @@ public Broadcast StartBroadcast(string sessionId, Boolean hls = true, List<Rtmp>
         /// including the broadcast ID.
         /// </returns>
         public Broadcast StopBroadcast(string broadcastId)
+        {
+            try
+            {
+                var task = StopBroadcastAsync(broadcastId);
+                task.Wait();
```
Comment (slorello89) [inline OpenTok/OpenTok.cs]: Still using Task.Wait()

File: OpenTok/OpenTok.cs [hunk 6]
```diff
@@ -481,6 +919,46 @@ public void ForceDisconnect(string sessionId, string connectionId)
         /// <param name="layout"></param>
         /// <returns>The Broadcast object. This object includes properties defining the archive, including the archive ID.</returns>
         public Broadcast StartBroadcast(string sessionId, Boolean hls = true, List<Rtmp> rtmpList = null, string resolution = null, int maxDuration = 7200, BroadcastLayout layout = null)
+        {
+            try
+            {
+                var task = StartBroadcastAsync(sessionId, hls, rtmpList, resolution, maxDuration, layout);
+                task.Wait();
```
Comment (slorello89) [inline OpenTok/OpenTok.cs]: Still using Task.Wait()

File: OpenTok/OpenTok.cs [hunk 7]
```diff
@@ -420,13 +820,33 @@ public Stream GetStream(string sessionId, string streamId)
         /// <param name="sessionId">The session ID corresponding to the session.</param>
         /// <returns>A List of <see cref="Stream"/> objects.</returns>
         public StreamList ListStreams(string sessionId)
+        {
+            try
+            {
+                var task = ListStreamsAsync(sessionId);
+                task.Wait();
```
Comment (slorello89) [inline OpenTok/OpenTok.cs]: Still using Task.Wait()

File: OpenTok/OpenTok.cs [hunk 8]
```diff
@@ -399,14 +779,34 @@ public void DeleteArchive(string archiveId)
         /// <param name="streamId">The stream ID.</param>
         /// <returns>The <see cref="Stream"/> object.</returns>
         public Stream GetStream(string sessionId, string streamId)
+        {
+            try
+            {
+                var task = GetStreamAsync(sessionId, streamId);
+                task.Wait();
```
Comment (slorello89) [inline OpenTok/OpenTok.cs]: Still using Task.Wait()

File: OpenTok/OpenTok.cs [hunk 9]
```diff
@@ -369,10 +708,29 @@ public ArchiveList ListArchives(int offset = 0 , int count = 0, string sessionId
         /// <param name="archiveId">The archive ID.</param>
         /// <returns>The <see cref="Archive"/> object.</returns>
         public Archive GetArchive(string archiveId)
+        {
+            try
+            {
+                var task = GetArchiveAsync(archiveId);
+                task.Wait();
```
Comment (slorello89) [inline OpenTok/OpenTok.cs]: Still using Task.Wait()

File: OpenTok/OpenTok.cs [hunk 10]
```diff
@@ -337,7 +650,33 @@ public Archive StopArchive(string archiveId)
         /// The number of archives to be returned. The maximum number of archives returned is 1000.
         /// </param>
         /// <returns>A List of <see cref="Archive"/> objects.</returns>
-        public ArchiveList ListArchives(int offset = 0 , int count = 0, string sessionId = "")
+        public ArchiveList ListArchives(int offset = 0, int count = 0, string sessionId = "")
+        {
+            try
+            {
+                var task = ListArchivesAsync(offset, count, sessionId);
+                task.Wait();
```
Comment (slorello89) [inline OpenTok/OpenTok.cs]: Still using Task.Wait()

File: OpenTok/OpenTok.cs [hunk 11]
```diff
@@ -317,14 +557,87 @@ public Archive StartArchive(string sessionId, string name = "", bool hasVideo =
...
+        /// <a href="https://tokbox.com/developer/guides/archiving/layout-control.html#dynamically-changing-the-layout-type-during-an-archive-recording">Configuring the video layout for OpenTok archive</a>.
+        /// </summary>
+        /// <param name="archiveId">The broadcast ID of the broadcasting session.</param>
+        /// <param name="layout">The BroadcastLayout that defines layout options for the broadcast.</param>
+        public void SetArchiveLayout(string archiveId, ArchiveLayout layout)
+        {
+            try
+            {
+                var task = SetArchiveLayoutAsync(archiveId, layout);
+                task.Wait();
```
Comment (slorello89) [inline OpenTok/OpenTok.cs]: Still using Task.Wait()

File: OpenTok/OpenTok.cs [hunk 12]
```diff
@@ -317,14 +557,87 @@ public Archive StartArchive(string sessionId, string name = "", bool hasVideo =
         /// <param name="archiveId">The archive ID of the archive you want to stop recording.</param>
         /// <returns>The Archive object corresponding to the archive being STOPPED.</returns>
         public Archive StopArchive(string archiveId)
+        {
+            try
+            {
+                var task = StopArchiveAsync(archiveId);
+                task.Wait();
```
Comment (slorello89) [inline OpenTok/OpenTok.cs]: Still using Task.Wait()

File: OpenTok/OpenTok.cs [hunk 13]
```diff
@@ -276,6 +404,118 @@ public string GenerateToken(string sessionId, Role role = Role.PUBLISHER, double
         /// The Archive object. This object includes properties defining the archive, including the archive ID.
         /// </returns>
         public Archive StartArchive(string sessionId, string name = "", bool hasVideo = true, bool hasAudio = true, OutputMode outputMode = OutputMode.COMPOSED, string resolution = null, ArchiveLayout layout = null)
+        {
+            try
+            {
+                var task = StartArchiveAsync(sessionId, name, hasVideo, hasAudio, outputMode, resolution, layout);
+                task.Wait();
```
Comment (slorello89) [inline OpenTok/OpenTok.cs]: Still using Task.Wait()

File: OpenTok/OpenTok.cs [hunk 14]
```diff
@@ -142,6 +143,84 @@ public OpenTok(int apiKey, string apiSecret, string apiUrl)
         /// method (to initialize an OpenTok session).
         /// </returns>
         public Session CreateSession(string location = "", MediaMode mediaMode = MediaMode.RELAYED, ArchiveMode archiveMode = ArchiveMode.MANUAL)
+        {
+            try
+            {
+                var task = CreateSessionAsync(location, mediaMode, archiveMode);
+                task.Wait();
```
Comment (slorello89) [inline OpenTok/OpenTok.cs]: Still using Task.Wait()

**Review by slorello89 — CHANGES_REQUESTED** [review]

**Comment by matt-lethargic** [comment]
Hi @gleborgne thank you for the PR, unfortunately, this is now out of date and seems like the comments were not addressed. SetArchiveLayout has been implemented all be it non-async. Are you happy to close this PR and I'll create a ticket for me to move over the remaining non-async methods sometime this week or next?

File: OpenTok/OpenTok.cs [hunk 15]
```diff
@@ -318,26 +584,87 @@ public Archive StartArchive(string sessionId, string name = "", bool hasVideo =
...
+                Url = url,
+                Headers = headers,
+                Data = data
+            };
+        }
 
         /// <summary>
-        /// Returns a List of <see cref="Archive"/> objects, representing archives that are both
-        /// both completed and in-progress, for your API key.
+        /// Sets the layout type for the broadcast. For a description of layout types, see
```
Comment (jeffswartz) [inline OpenTok/OpenTok.cs:641]:
```suggestion
        /// Sets the layout type for the archive. For a description of layout types, see
```

File: OpenTok/OpenTok.cs [hunk 16]
```diff
@@ -318,26 +584,87 @@ public Archive StartArchive(string sessionId, string name = "", bool hasVideo =
...
-        /// <param name="offset">
-        /// The index offset of the first archive. 0 is offset of the most recently started archive.
-        /// 1 is the offset of the archive that started prior to the most recent archive.
-        /// </param>
-        /// <param name="count">
-        /// The number of archives to be returned. The maximum number of archives returned is 1000.
-        /// </param>
-        /// <returns>A List of <see cref="Archive"/> objects.</returns>
-        public ArchiveList ListArchives(int offset = 0 , int count = 0, string sessionId = "")
+        /// <param name="archiveId">The broadcast ID of the broadcasting session.</param>
```
Comment (jeffswartz) [inline OpenTok/OpenTok.cs:644]:
```suggestion
        /// <param name="archiveId">The archive ID.</param>
```

File: OpenTok/OpenTok.cs [hunk 17]
```diff
@@ -318,26 +584,87 @@ public Archive StartArchive(string sessionId, string name = "", bool hasVideo =
...
+        /// <param name="layout">The BroadcastLayout that defines layout options for the broadcast.</param>
```
Comment (jeffswartz) [inline OpenTok/OpenTok.cs:645]:
```suggestion
        /// <param name="layout">The ArchiveLayout that defines layout options for the archive.</param>
```

(And make similar edits for the Async version of the method.

**Review by jeffswartz — CHANGES_REQUESTED** [review]

**Pull request closed without merging**